File: LeetCode/course_schedule.js
var canFinish = function (numCourses, arr) {
    const graph = {};
    let preReq = true;

    for (let i = 0; i < arr.length; i++) {
        if (!graph[arr[i][0]]) graph[arr[i][0]] = [];
        graph[arr[i][0]].push(arr[i][1]);
    }

    flagged = new Set();
    const memo = {};

    const graphArr = Object.keys(graph);
    for (let i = 0; i < graphArr.length; i++) {
        if (dfs(graph, graphArr[i], flagged, memo) === false) {
            return false;
        }
    }

    return preReq;
};

var dfs = function (graph, classSelect, flagged, memo = {}) {
    if (graph[classSelect] === undefined) return true;
    if (flagged.has(classSelect)) return false;
    if (classSelect in memo) { return memo[classSelect] }

    let pre = true;
    flagged.add(classSelect);

    for (let i = 0; i < graph[classSelect].length; i++) {
        let copy = new Set(flagged);
        const req = graph[classSelect][i];

        const test = dfs(graph, req, copy, memo);
        if (!test) return false;
    }
    memo[classSelect] = true;
    return memo[classSelect];
}

File: LeetCode/course_schedule.py
class Solution(object):
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        visited = [0 for _ in range(numCourses)]
        
        # create graph
        for pair in prerequisites:
            x, y = pair
            graph[x].append(y)
            
        # visit each node
        for course in range(numCourses):
            if not self.dfs(graph, visited, course): return False
            
        return True
    
    def dfs(self, graph, visited, course):
        # if ith node is marked as being visited, then a cycle is found
        if visited[course] == -1: return False
        
        # if it is done visted, then do not visit again
        if visited[course] == 1: return True
        
        # mark as being visited
        visited[course] = -1
        
        # visit all the neighbours
        for prereq in graph[course]:
            if not self.dfs(graph, visited, prereq): return False
            
        # after visit all the neighbours, mark it as done visited
        visited[course] = 1
        
        return True
        